File: src/cli/commands/address.zig
//! Address commands: to-checksum, compute-address, create2

const std = @import("std");
const primitives = @import("primitives");
const crypto_mod = @import("crypto");
const cli = @import("../mod.zig");

const Context = cli.Context;
const CliError = cli.CliError;
const Address = primitives.Address;

/// Convert address to checksummed format (EIP-55)
/// Usage: chop to-checksum <address>
pub fn toChecksum(ctx: *Context, args: []const []const u8) CliError!void {
    if (args.len == 0) {
        try ctx.err("Usage: chop to-checksum <address>\n", .{});
        return CliError.MissingArgument;
    }

    const input = args[0];

    // Parse address
    const addr = Address.fromHex(input) catch {
        try ctx.err("error: invalid address '{s}'\n", .{input});
        return CliError.InvalidAddress;
    };

    // Get checksummed version
    const checksummed = addr.toChecksummed();

    if (ctx.format == .json) {
        try ctx.print("{{\"address\":\"{s}\"}}\n", .{&checksummed});
    } else {
        try ctx.print("{s}\n", .{&checksummed});
    }
}

/// Compute contract address from deployer and nonce (CREATE)
/// Usage: chop compute-address <deployer> <nonce>
pub fn computeAddress(ctx: *Context, args: []const []const u8) CliError!void {
    if (args.len < 2) {
        try ctx.err("Usage: chop compute-address <deployer> <nonce>\n", .{});
        return CliError.MissingArgument;
    }

    const deployer_str = args[0];
    const nonce_str = args[1];

    // Parse deployer address
    const deployer = Address.fromHex(deployer_str) catch {
        try ctx.err("error: invalid deployer address '{s}'\n", .{deployer_str});
        return CliError.InvalidAddress;
    };

    // Parse nonce
    const nonce = std.fmt.parseUnsigned(u64, nonce_str, 10) catch {
        try ctx.err("error: invalid nonce '{s}'\n", .{nonce_str});
        return CliError.InvalidArgument;
    };

    // Compute contract address
    const contract_addr = Address.getContractAddress(ctx.allocator, deployer, nonce) catch {
        try ctx.err("error: failed to compute contract address\n", .{});
        return CliError.EncodingError;
    };
    const checksummed = contract_addr.toChecksummed();

    if (ctx.format == .json) {
        try ctx.print("{{\"address\":\"{s}\"}}\n", .{&checksummed});
    } else {
        try ctx.print("{s}\n", .{&checksummed});
    }
}

/// Compute CREATE2 address
/// Usage: chop create2 <deployer> <salt> <init_code_hash>
pub fn create2(ctx: *Context, args: []const []const u8) CliError!void {
    if (args.len < 3) {
        try ctx.err("Usage: chop create2 <deployer> <salt> <init_code_hash>\n", .{});
        return CliError.MissingArgument;
    }

    const deployer_str = args[0];
    const salt_str = args[1];
    const init_code_hash_str = args[2];

    // Parse deployer address
    const deployer = Address.fromHex(deployer_str) catch {
        try ctx.err("error: invalid deployer address '{s}'\n", .{deployer_str});
        return CliError.InvalidAddress;
    };

    // Parse salt (32 bytes)
    const salt = parseBytes32(salt_str) catch {
        try ctx.err("error: invalid salt '{s}'\n", .{salt_str});
        return CliError.InvalidHex;
    };

    // Parse init code hash (32 bytes)
    const init_code_hash = parseBytes32(init_code_hash_str) catch {
        try ctx.err("error: invalid init code hash '{s}'\n", .{init_code_hash_str});
        return CliError.InvalidHex;
    };

    // Compute CREATE2 address
    const contract_addr = Address.getCreate2Address(deployer, salt, init_code_hash);
    const checksummed = contract_addr.toChecksummed();

    if (ctx.format == .json) {
        try ctx.print("{{\"address\":\"{s}\"}}\n", .{&checksummed});
    } else {
        try ctx.print("{s}\n", .{&checksummed});
    }
}

fn parseBytes32(input: []const u8) ![32]u8 {
    const hex = if (input.len >= 2 and input[0] == '0' and (input[1] == 'x' or input[1] == 'X'))
        input[2..]
    else
        input;

    if (hex.len != 64) return error.InvalidLength;

    var result: [32]u8 = undefined;
    for (0..32) |i| {
        result[i] = std.fmt.parseUnsigned(u8, hex[i * 2 .. i * 2 + 2], 16) catch return error.InvalidHex;
    }
    return result;
}

File: src/cli/commands/address.ts
/**
 * Address utility CLI commands.
 *
 * Commands:
 * - to-check-sum-address: Convert address to EIP-55 checksummed form
 * - compute-address: Compute CREATE address from deployer + nonce
 * - create2: Compute CREATE2 address from deployer + salt + init-code
 */

import { Args, Command, Options } from "@effect/cli"
import { Console, Data, Effect } from "effect"
import { Address, Hash, Keccak256 } from "voltaire-effect"
import { handleCommandErrors, jsonOption, validateHexData } from "../shared.js"

// ============================================================================
// Error Types
// ============================================================================

/** Error for invalid Ethereum addresses */
export class InvalidAddressError extends Data.TaggedError("InvalidAddressError")<{
	readonly message: string
	readonly address: string
}> {}

/** Error for invalid hex data (salt, init-code) */
export class InvalidHexError extends Data.TaggedError("InvalidHexError")<{
	readonly message: string
	readonly hex: string
}> {}

/** Error for address computation failures */
export class ComputeAddressError extends Data.TaggedError("ComputeAddressError")<{
	readonly message: string
	readonly cause?: unknown
}> {}

// ============================================================================
// Validation Helpers
// ============================================================================

/** Validate and parse an address from a hex string */
const validateAddress = (raw: string) =>
	Address.fromHex(raw).pipe(
		Effect.catchAll(() =>
			Effect.fail(
				new InvalidAddressError({
					message: `Invalid address: "${raw}". Expected a 40-character hex string with 0x prefix.`,
					address: raw,
				}),
			),
		),
	)

/** Validate hex data and convert to Uint8Array */
const validateHexDataAsInvalidHex = (raw: string): Effect.Effect<Uint8Array, InvalidHexError> =>
	validateHexData(raw, (message, hex) => new InvalidHexError({ message, hex }))

/** Validate a 32-byte salt from hex string */
const validateSalt = (raw: string) =>
	Effect.gen(function* () {
		if (!raw.startsWith("0x")) {
			return yield* Effect.fail(
				new InvalidHexError({
					message: `Invalid salt: "${raw}". Expected a 32-byte (64-character) hex string with 0x prefix.`,
					hex: raw,
				}),
			)
		}
		return yield* Hash.fromHex(raw).pipe(
			Effect.catchAll(() =>
				Effect.fail(
					new InvalidHexError({
						message: `Invalid salt: "${raw}". Expected a 32-byte (64-character) hex string with 0x prefix.`,
						hex: raw,
					}),
				),
			),
		)
	})

// ============================================================================
// Handler Logic (testable, separated from CLI wiring)
// ============================================================================

/** Core to-check-sum-address logic: validates and checksums an address. */
export const toCheckSumAddressHandler = (rawAddr: string) =>
	Effect.gen(function* () {
		const addr = yield* validateAddress(rawAddr)
		return yield* Address.toChecksummed(addr)
	})

/** Core compute-address logic: computes CREATE address from deployer + nonce. */
export const computeAddressHandler = (rawDeployer: string, rawNonce: string) =>
	Effect.gen(function* () {
		const deployer = yield* validateAddress(rawDeployer)

		const nonce = yield* Effect.try({
			try: () => {
				const n = BigInt(rawNonce)
				if (n < 0n) {
					throw new Error("Nonce must be non-negative")
				}
				return n
			},
			catch: (e) =>
				new ComputeAddressError({
					message: `Invalid nonce: "${rawNonce}". ${e instanceof Error ? e.message : "Expected a non-negative integer."}`,
					cause: e,
				}),
		})

		const contractAddr = yield* Address.calculateCreateAddress(deployer, nonce).pipe(
			Effect.catchAll((e) =>
				Effect.fail(
					new ComputeAddressError({
						message: `Failed to compute CREATE address: ${e instanceof Error ? e.message : String(e)}`,
						cause: e,
					}),
				),
			),
		)

		return yield* Address.toChecksummed(contractAddr)
	})

/** Core create2 logic: computes CREATE2 address from deployer + salt + init-code. */
export const create2Handler = (rawDeployer: string, rawSalt: string, rawInitCode: string) =>
	Effect.gen(function* () {
		const deployer = yield* validateAddress(rawDeployer)
		const salt = yield* validateSalt(rawSalt)
		const initCode = yield* validateHexDataAsInvalidHex(rawInitCode)

		const contractAddr = yield* Address.calculateCreate2Address(deployer, salt, initCode).pipe(
			Effect.catchAll((e) =>
				Effect.fail(
					new ComputeAddressError({
						message: `Failed to compute CREATE2 address: ${e instanceof Error ? e.message : String(e)}`,
						cause: e,
					}),
				),
			),
		)

		return yield* Address.toChecksummed(contractAddr)
	})

// ============================================================================
// Commands
// ============================================================================

/**
 * `chop to-check-sum-address <addr>`
 *
 * Convert an Ethereum address to its EIP-55 checksummed form.
 */
export const toCheckSumAddressCommand = Command.make(
	"to-check-sum-address",
	{
		addr: Args.text({ name: "addr" }).pipe(Args.withDescription("Ethereum address to checksum")),
		json: jsonOption,
	},
	({ addr, json }) =>
		Effect.gen(function* () {
			const result = yield* toCheckSumAddressHandler(addr)
			if (json) {
				yield* Console.log(JSON.stringify({ result }))
			} else {
				yield* Console.log(result)
			}
		}).pipe(Effect.provide(Keccak256.KeccakLive), handleCommandErrors),
).pipe(Command.withDescription("Convert address to EIP-55 checksummed form"))

/**
 * `chop compute-address --deployer <addr> --nonce <n>`
 *
 * Compute the contract address that would be deployed via CREATE.
 */
export const computeAddressCommand = Command.make(
	"compute-address",
	{
		deployer: Options.text("deployer").pipe(Options.withDescription("Deployer address")),
		nonce: Options.text("nonce").pipe(Options.withDescription("Transaction nonce")),
		json: jsonOption,
	},
	({ deployer, nonce, json }) =>
		Effect.gen(function* () {
			const result = yield* computeAddressHandler(deployer, nonce)
			if (json) {
				yield* Console.log(JSON.stringify({ result }))
			} else {
				yield* Console.log(result)
			}
		}).pipe(Effect.provide(Keccak256.KeccakLive), handleCommandErrors),
).pipe(Command.withDescription("Compute CREATE contract address from deployer + nonce"))

/**
 * `chop create2 --deployer <addr> --salt <hex> --init-code <hex>`
 *
 * Compute the contract address that would be deployed via CREATE2.
 */
export const create2Command = Command.make(
	"create2",
	{
		deployer: Options.text("deployer").pipe(Options.withDescription("Deployer/factory address")),
		salt: Options.text("salt").pipe(Options.withDescription("32-byte salt as hex")),
		initCode: Options.text("init-code").pipe(Options.withDescription("Contract init code as hex")),
		json: jsonOption,
	},
	({ deployer, salt, initCode, json }) =>
		Effect.gen(function* () {
			const result = yield* create2Handler(deployer, salt, initCode)
			if (json) {
				yield* Console.log(JSON.stringify({ result }))
			} else {
				yield* Console.log(result)
			}
		}).pipe(Effect.provide(Keccak256.KeccakLive), handleCommandErrors),
).pipe(Command.withDescription("Compute CREATE2 contract address"))

// ============================================================================
// Exports
// ============================================================================

/** All address-related subcommands for registration with the root command. */
export const addressCommands = [toCheckSumAddressCommand, computeAddressCommand, create2Command] as const
